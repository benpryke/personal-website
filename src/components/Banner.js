import React from 'react';
import classNames from 'classnames';

/**
 * A Banner is a full-width segment spanning the page
 */
export default class Banner extends React.Component {
  static defaultProps = {
    fadeIn: true,
    fadeInOffset: 100,
  };
  
  state = {
    fadedIn: false,
  };

  constructor(props) {
    super(props);

    if (props.fadeIn) {
      this.ref = React.createRef();
      this.triggerFadeIn = this.triggerFadeIn.bind(this);
      this.scrollHandler = document.addEventListener('scroll', this.triggerFadeIn);
    }
  }
  
  componentDidMount() {
    if (this.props.fadeIn) {
      this.triggerFadeIn();
    }
  }

  triggerFadeIn() {
    const { fadeIn, fadeInOffset } = this.props;

    // Only trigger the fade in animation if it's switched on
    if (!fadeIn) return;
    
    // To make the fade-in more visible, we start at an offset.
    // Note: if the banner height is less than the offset and at the bottom of
    // the page, we will never be able to trigger the animation.
    const bannerTop = this.ref.current ? this.ref.current.offsetTop : Infinity;
    const windowBottom = window.scrollY + window.innerHeight;
    const fadedIn = windowBottom >= bannerTop + fadeInOffset;

    if (fadedIn) {
      document.removeEventListener('scroll', this.scrollHandler);
      this.setState({ fadedIn });
    }
  }

  render() {
    const { className, children, fadeIn } = this.props;
    const { fadedIn } = this.state;
    let { style } = this.props;

    // If `fadeIn` is true, we start invisible
    if (fadeIn && !fadedIn) {
      style = style || {};
      style['opacity'] = 0;
    }

    return (
      <section
        ref={this.ref}
        className={classNames('banner', {
          [className]: className,
          'faded-in': fadedIn,
        })}
        style={style}
      >
        {children}
      </section>
    );
  }
}
