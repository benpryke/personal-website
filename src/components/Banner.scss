.banner {
  position: relative;
  min-width: $min-content-width;
  max-width: $max-content-width;
  min-height: $min-section-height;
  margin: auto;
  padding: $section-vertical-padding 10%;
  text-align: center;
  box-sizing: border-box;
  
  &>h2 {
    margin: 0 0 30px;
  }
  
  &>p {
    color: $primary-text-colour;
  }
}

.faded-out {
  opacity: 0;
}

.faded-in {
  animation: fade-in 0.5s ease-out;
}

// Fade-in animation
@keyframes fade-in {
  0% {
    opacity: 0;
    top: 100px;
  }

  100% {
    opacity: 1;
    top: 0;
  }
}

@media only screen and (min-width: $screen-width-s) {
  .banner {
    box-sizing: inherit;
    
    &:not(.showcase):not(.logos) {
      padding: $section-vertical-padding 20%;
    }
  }
}

// 20x + 1500 + 20x = 100x => 100x = 2500
@media only screen and (min-width: $screen-width-xl) {
  .banner {
    padding: $section-vertical-padding calc((100% - #{$max-content-width}) / 2);
  }
}
