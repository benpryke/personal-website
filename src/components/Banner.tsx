import React, { useState, useEffect, useRef } from "react";
import classNames from "classnames";

export interface BannerProps {
  /** CSS class name for the Banner */
  className?: string;
  /** Whether or not to fade the Banner in as it's scrolled into view */
  fadeIn?: boolean;
  /**
   * How many pixels above the bottom of the viewport the Banner has to be before it fades in.
   * This ensures that the animation is visible.
   */
  fadeInOffset?: number;
  /** CSS style */
  style?: React.CSSProperties;
  /** Banners must have children */
  children?: React.ReactNode;
}

const Banner: React.FC<BannerProps> = ({
  fadeIn = true,
  fadeInOffset = 100,
  className,
  style,
  children,
}) => {
  const [fadedIn, setFadedIn] = useState(false);
  const [fadedOut, setFadedOut] = useState(false);
  const ref = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (fadeIn) {
      const triggerFadeIn = () => {
        if (ref.current) {
          const bannerTop = ref.current.offsetTop;
          const windowBottom = window.scrollY + window.innerHeight;
          const fadedIn = windowBottom >= bannerTop + fadeInOffset;

          if (fadedIn) {
            document.removeEventListener("scroll", triggerFadeIn);
            setFadedIn(true);
          } else if (!fadedOut) {
            setFadedOut(true);
          }
        }
      };

      triggerFadeIn();
      document.addEventListener("scroll", triggerFadeIn);
      return () => document.removeEventListener("scroll", triggerFadeIn);
    } else {
      setFadedIn(true);
    }
  }, [fadeIn, fadeInOffset, fadedOut]);

  return (
    <section
      ref={ref}
      className={classNames("banner", {
        [className ?? ""]: className,
        "faded-out": fadedOut && !fadedIn,
        "faded-in": fadeIn && fadedIn,
      })}
      style={style}
    >
      {children}
    </section>
  );
};

export default Banner;
