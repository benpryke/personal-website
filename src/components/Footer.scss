.footer {
  max-width: initial;
  min-height: 0;
  padding-top: 80px !important;
  padding-bottom: 120px !important;
  background: $background-colour;
}
