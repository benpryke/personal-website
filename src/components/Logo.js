import React from 'react';

/**
 * A logo for the Logos component
 */
export default class Logo extends React.Component {
  render() {
    const { name, src, url } = this.props;

    return (
      <a className='logo' href={url} key={name}>
        <img src={src} alt={name}/>
      </a>
    );
  }
}
