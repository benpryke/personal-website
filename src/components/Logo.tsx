import React from "react";

import { LazyLoad } from "./LazyLoad";

export interface LogoProps {
  /** Name of the organisation or entity */
  name: string;
  /** Logo URL */
  src: string;
  /** URL to their website or a relevant page */
  url: string;
}

/**
 * A logo for the Logos component
 */
export const Logo: React.FC<LogoProps> = ({ name, src, url }) => {
  return (
    <a className="logo" href={url} key={name}>
      <LazyLoad offset={100}>
        <img src={src} alt={name} />
      </LazyLoad>
    </a>
  );
};
