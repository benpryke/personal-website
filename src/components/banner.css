.banner {
  position: relative;
  min-width: var(--min-content-width);
  max-width: var(--max-content-width);
  min-height: var(--min-section-height);
  margin: auto;
  padding: var(--section-vertical-padding) 10%;
  text-align: center;
  box-sizing: border-box;

  & > h2 {
    margin: 0 0 30px;
  }

  & > p {
    color: var(--primary-text-colour);
  }
}

.faded-out {
  opacity: 0;
}

.faded-in {
  animation: fade-in 0.5s ease-out;
}

/* Fade-in animation */
@keyframes fade-in {
  0% {
    opacity: 0;
    top: 100px;
  }

  100% {
    opacity: 1;
    top: 0;
  }
}

@media only screen and (--screen-width-s) {
  .banner {
    box-sizing: inherit;
  }
}

/* 20x + 1500 + 20x = 100x => 100x = 2500 */
@media only screen and (--screen-width-l) {
  .banner {
    padding: var(--section-vertical-padding)
      calc((100% - var(--max-content-width)) / 2);
  }
}
