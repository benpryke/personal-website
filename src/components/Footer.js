import React, { Component } from 'react';

import Banner from './Banner';
import { LinkedInButton, TwitterButton, GitHubButton, StackOverflowButton, InstagramButton } from './SocialButton';

/**
 * Page footer
 */
export default class Footer extends Component {
  render() {
    return (
      <Banner className='footer' {...this.props}>
        <h2>Get in touch</h2>
        <footer>
          <LinkedInButton/>
          <TwitterButton/>
          <GitHubButton/>
          <StackOverflowButton/>
          <InstagramButton/>
        </footer>
      </Banner>
    );
  }
}
