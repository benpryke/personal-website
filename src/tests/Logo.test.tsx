import pretty from "pretty";
import React from "react";

import { render } from "@testing-library/react";

import { Logo, type LogoProps } from "../components/Logo";

const props: LogoProps = {
  name: "name",
  src: "src",
  url: "url",
};

describe("Logo", () => {
  it("renders without crashing", () => {
    render(<Logo {...props} />);
  });

  it("should match the snapshot", () => {
    const { container } = render(<Logo {...props} />);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
