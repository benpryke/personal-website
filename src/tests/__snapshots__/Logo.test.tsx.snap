// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Logo > should match the snapshot 1`] = `
"<a class="logo" href="url">
  <div></div>
</a>"
`;
